crates/mpc-protocols/src/mpc_service.rs

```diff
@@ -0,0 +1,116 @@
+use crate::mpc_events::{CreatedProofMPCEvent, MPCEvent};
+use std::collections::{HashMap, VecDeque};
+use sui_types::base_types::ObjectID;
+use sui_types::event::Event;
+use tokio::sync::{mpsc, Mutex, RwLock};
+
+const MAX_ACTIVE_MPC_INSTANCES: usize = 100;
+
+pub enum MPCInput {
+    InitEvent(CreatedProofMPCEvent),
+    Message,
+    OutputEvent,
+}
+
+struct MPCInstance {
+    status: MPCStatus,
+    /// The channel to send messages to active MPC instance
+    /// Every new message related to this instance will be forwarded to this channel and will be handled synchronously
+    messages_handler_sender: Option<mpsc::Sender<MPCInput>>,
+    pending_messages: Vec<MPCInput>,
+}
+
+/// Possible status of an MPC session
+/// - Active: The session is currently running, new messages will be forwarded to the session
+/// - Pending: The session is waiting for a slot to become active, when the init received there were already `MAX_ACTIVE_MPC_INSTANCES` active sessions
+/// - Finished: The session has finished, no more messages will be forwarded. The session will be removed from the service after a timeout.
+/// We want to keep it for some time so leftover messages related to it won't be treated as malicious.
+#[derive(Clone, Copy)]
+enum MPCStatus {
+    Active,
+    Pending,
+    Finished,
+}
+
+/// The `MPCService` is responsible for managing MPC instances:
+/// - It keeps track of all MPC instances
+/// - Runs the MPC session for each active instance
+/// - Ensures that the number of active sessions does not go over `MAX_ACTIVE_MPC_INSTANCES` at the same time
+pub struct MPCService {
+    mpc_instances: Mutex<HashMap<ObjectID, MPCInstance>>,
+    pending: Mutex<VecDeque<ObjectID>>,
+    active_instances_counter: Mutex<usize>,
+}
+
+impl MPCService {
+    pub fn new() -> Self {
+        Self {
+            mpc_instances: Mutex::new(HashMap::new()),
+            pending: Mutex::new(VecDeque::new()),
+            active_instances_counter: Mutex::new(0),
+        }
+    }
+
+    /// Spawns an asynchronous task to handle incoming messages for a new MPC instance.
+    /// The [`MPCService`] will forward any message related to that instance to this channel.
+    fn spawn_mpc_messages_handler(&self, mut receiver: mpsc::Receiver<MPCInput>) {
+        tokio::spawn(async move {
+            while let Some(message) = receiver.recv().await {
+                // TODO (#235): Implement MPC messages handling
+            }
+        });
+    }
+
+    /// Filter the relevant MPC events from all the consensus output transactions events & handle them
+    pub fn handle_mpc_events(&self, events: &Vec<Event>) -> anyhow::Result<()> {
```

Comment:
Re: [this comment](https://github.com/dwallet-labs/dwallet-network/pull/234/files#r1763481145), the doc here says that `handle_mpc_events()` filters the events from _all the consensus output transactions_. So the way you're using this function in `authority.rs` is in contradiction with this doc; which is the correct one?

Reply:
This comment is wrong, we changed it. Lmk what you think.

Reply:
The new doc (as seen in #244) looks good to me.